src/test/groovy/pl/allegro/tech/opel/OpelEngineFunctionIntegrationSpec.groovy

```diff
@@ -0,0 +1,207 @@
+                """
+                    val x = 3;
+                    val f = (a, b) -> {
+                        val y = a*x*x;
+                        y + b*x
+                    };
+                    f(1,3)"""
+        ]
+        expResult << [10, 10, 18]
+    }
+
+    @Unroll
+    def 'should create higher-order function and call it'() {
+        given:
+        def engine = create().build()
+
+        expect:
+        engine.eval(input).get() == expResult
+
+        where:
+        input << ["""
+                    val f = a -> {
+                        b -> {a*2+b}
+                    };
+                    f(2)(3)
+                """,
+                """
+                    val f = a -> {
+                        b -> {
+                            c -> a*a*a + b*b + c
+                        }
+                    };
+                    f(1)(2)(3)
+                """,
+        ]
+        expResult << [7, 8]
+    }
+
+    @Unroll
+    def 'should return function in #input'() {
+        given:
+        def engine = create().build()
+
+        when:
+        def function = engine.eval(input).get()
+
+        then:
+        function instanceof OpelAsyncFunction
+        function.apply(args).get() == expResult
+
+        where:
+        input                                                    || args                                     | expResult
+        "() -> 3"                                                || []                                       | 3
+        "val a = 2; x -> a * x"                                  || [completedFuture(2)]                     | 4
+        "x -> x * 3"                                             || [completedFuture(3)]                     | 9
+        "(x) -> x * 3"                                           || [completedFuture(3)]                     | 9
+        "val f = x -> x * 3; f"                                  || [completedFuture(4)]                     | 12
+        "val f = (x, y) -> x*x + y*y*y; f"                       || [completedFuture(1), completedFuture(2)] | 9
+        "val f = (x, y) -> x*x + y*y*y; val g = x -> f(x, 2); g" || [completedFuture(3)]                     | 17
+    }
+
+    def 'should create higher-order function and return inner function'() {
+        given:
+        def engine = create().build()
+        def input = """
+                    val f = a -> {
+                        b -> {a*2+b}
+                    };
+                    f(2)
+                """
+
+        when:
+        def function = engine.eval(input).get()
+
+        then:
+        function instanceof OpelAsyncFunction
+        function.apply([completedFuture(3)]).get() == 7
+    }
+
+    @Unroll
+    def "should failed when missing arguments (#input)"() {
+        given:
+        def engine = create().build()
+
+        when:
+        engine.eval(input).get()
+
+        then:
+        def ex = thrown ExecutionException
+        ex.cause instanceof OpelException
```

Comment:
I don't like the fact that we only check if `OpelException` was thrown while cause is completely ignored. I can modify the first case to `val fun = x -> x*x; fun(4) * 'a'` and this test will pass. Maybe it's a good time to introduce hierarchy of exceptions or check for sufficient messages in tests.